### PR TITLE
OpenRouter interleaved reasoning forwarding for tool-call continuations

### PR DESCRIPTION
**Pull request by ZhenyaPav**
Disclosure - the changes were vibe coded.

**Description:**  
This PR implements interleaved reasoning on OpenRouter for models beyond Gemini, and adds a mode selector.

**Before:**  
- Interleaved reasoning replay on OpenRouter was incomplete/inconsistent for non-Gemini models.  
- Tool continuation requests often lacked usable forwarded reasoning metadata (`reasoning_details`/fallback reasoning), especially in some streaming paths.  
- Result: for most models, reasoning started over after every tool call.

**What changed:**  
- Captures OpenRouter reasoning more robustly (streaming + non-stream): `reasoning`, `reasoning_content`, and `reasoning_details`.  
- Replays reasoning on assistant tool-call messages by:
  - preferring `reasoning_details` signatures
  - falling back to plaintext `reasoning` when needed  
- Adds configurable replay scope:
  - `Disabled`: never resend prior tool-turn reasoning to the model after tool calls (Default option).
  - `Since Last User Message`: resend reasoning for any tool turns that happened after the latest user message (Works fine with Kimi, but returns an error on Deepseek 3.2).
  - `Active Tool Chain`: resend reasoning only while still inside the current unresolved tool loop; once a normal assistant reply is produced, old tool-chain reasoning is no longer resent (Recommended).
- If the user switches to a different model/provider, the old saved reasoning from the previous model will not be sent into the new request.

**Models I have tested  my changes on (OpenRouter):**  
- DeepSeek 3.2  
- Kimi 2.5  
- GLM 4.7  
- GLM 4.7 Flash  

Control check:
- DeepSeek 3.1 Terminus Exacto behaves as expected as a no-reasoning control model.

## Checklist:

- [X] I have read the [Contribution guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ZhenyaPav** [comment]
One open issue I found during testing (DeepSeek 3.2 via OpenRouter): some requests return `400` when the assistant text and tool call are split into two consecutive assistant messages:

1. assistant text-only message  
2. assistant tool-call message (`tool_calls`, reasoning)  
3. tool result message

ST currently stores tool calls as a separate assistant message, and my interleaved-reasoning forwarding follows that structure. This works for many models/providers, but DeepSeek 3.2 seems stricter and may expect a single assistant turn containing `content + tool_calls (+ reasoning)` before the `tool` response.

Likely solution: keep ST internal history unchanged, but add provider-specific outbound serialization that merges adjacent assistant text + assistant tool-call into one assistant message for request payloads (DeepSeek/OpenRouter path), preserving `tool_call_id` mapping and reasoning forwarding semantics.

I need advise, should I implement the merge solution only for DeepSeek/OpenRouter, for all OpenRouter models, or do something different?

**Comment by Cohee1207** [comment]
Pls no provider-specific stuff. I'd first propose to reach out to OpenRouter and ask them to fix it on their end.

**Comment by ZhenyaPav** [comment]
> Pls no provider-specific stuff. I'd first propose to reach out to OpenRouter and ask them to fix it on their end.

The issue I described in the comment is specific to deepseek, and not OpenRouter.
https://forum.cursor.com/t/deepseek-reasoner-via-its-api-now-always-fails-with-deepseek-reasoner-does-not-support-successive-user-or-assistant-messages/92864

Either way, I think it should be fixed separately from this PR, most likely by changes to the existing custom prompt post processing, or adding a new option (this is exactly the case of successive assistant messages).

File: public/index.html [hunk 1]
```diff
@@ -4474,6 +4474,16 @@ <h4 class="standoutHeader">
                                         <small data-i18n="Max">Max</small>
                                     </div>
                                 </div>
+                                <div class="flex-container alignItemsBaseline">
```
Comment (Cohee1207) [inline public/index.html]: I think it should be under response configuration near the tool calling toggle and not in the Advanced Formatting. Also scoped for openrouter and hidden for other sources.

One more reason is that this only concerns Chat Completion as TC doesn't have tool calling.

File: public/scripts/openai.js [hunk 2]
```diff
@@ -6000,6 +6051,24 @@ export function isAudioInliningSupported() {
     }
 }
 
+/**
+ * Gets the OpenRouter tool-call reasoning forwarding mode.
+ * @returns {'disabled'|'since_last_user'|'active_chain'} Reasoning forwarding mode
+ */
+function getOpenRouterToolReasoningMode() {
+    const mode = String(power_user.reasoning?.forward_tool_chains_mode ?? '');
+    if (mode === 'disabled' || mode === 'since_last_user' || mode === 'active_chain') {
```
Comment (Cohee1207) [inline public/scripts/openai.js]: Add a enum / const object to remove hard coded string constants and simplify these checks to `Object.values.includes`.

File: public/scripts/openai.js [hunk 3]
```diff
@@ -6000,6 +6051,24 @@ export function isAudioInliningSupported() {
...
+        return mode;
+    }
+
+    // Legacy boolean fallback.
```
Comment (Cohee1207) [inline public/scripts/openai.js]: There can't be any legacy because it's 100% new feature

File: public/scripts/power-user.js [hunk 4]
```diff
@@ -282,6 +282,8 @@ export const power_user = {
         name: DEFAULT_REASONING_TEMPLATE,
         auto_parse: false,
         add_to_prompts: false,
+        forward_tool_chains_mode: 'disabled',
```
Comment (Cohee1207) [inline public/scripts/power-user.js]: I'd move these to Chat Completion settings object. Would also make them presettable.

File: public/scripts/openai.js [hunk 5]
```diff
@@ -6008,9 +6077,9 @@ export function isAudioInliningSupported() {
 export function isReasoningSignatureSupported(settings = oai_settings) {
     // If it's Vertex AI or Makersuite, that's OK - convertGooglePrompt() will handle it later
     const isGoogle = [chat_completion_sources.VERTEXAI, chat_completion_sources.MAKERSUITE].includes(settings.chat_completion_source);
-    // Need a more crunchy check for OpenRouter: look for Gemini models
-    const isOpenRouterGemini = settings.chat_completion_source === chat_completion_sources.OPENROUTER && /google\/gemini/i.test(settings.openrouter_model);
-    return isGoogle || isOpenRouterGemini;
+    const isOpenRouter = settings.chat_completion_source === chat_completion_sources.OPENROUTER
```
Comment (Cohee1207) [inline public/scripts/openai.js]: Does this mean that with the default 'disabled' setting, OpenRouter for Gemini won't forward signatures unless enabled?

File: public/scripts/power-user.js [hunk 6]
```diff
@@ -282,6 +282,8 @@ export const power_user = {
...
+        forward_tool_chains: false,
```
Comment (Cohee1207) [inline public/scripts/power-user.js]: This new property looks unused?

File: public/scripts/openai.js [hunk 7]
```diff
@@ -940,10 +948,27 @@ async function populateChatHistory(messages, prompts, chatCompletion, type = nul
 
         if (canUseTools && Array.isArray(chatPrompt.invocations)) {
             /** @type {import('./tool-calling.js').ToolInvocation[]} */
```
Comment (Cohee1207) [inline public/scripts/openai.js]: Type comment is misaligned

**Review by Cohee1207 — CHANGES_REQUESTED** [review]

File: public/scripts/power-user.js (hunk 6 above)
Comment (ZhenyaPav) [reply public/scripts/power-user.js]: forward_tool_chains is indeed unused (I originally implemented it before moving to a dropdown). I will remove it

**Comment by Cohee1207** [comment]
Please let me know when it's fully ready for a final review and test. I'll try my best to verify all 3 options.

**Comment by ZhenyaPav** [comment]
Should be ready to merge now.

- I have decoupled Gemini signature forwarding from my changes. The dropdown I added should not change Gemini functionality, only plaintext reasoning.
- Moved the interleaved reasoning settings to Chat Completion panel, specific to OpenRouter
- Had to significantly rework "Active Tool Chain" option - it did not work properly before. The idea behind it is to stop interleaved thinking chain when LLM returns a message that contains both content and tool calls. I am not entirely sure if it will behave appropriately in every edge case, but it did not produce any errors during my testing.
- `forward_tool_chains_mode` (the boolean option I initially added) was removed.
- Translation was added to the new UI options. (Machine translation with an LLM)

File: public/scripts/openai.js (hunk 5 above)
Comment (ZhenyaPav) [reply public/scripts/openai.js]: Fixed in the most recent changes, my changes should no longer impact Gemini functionality

File: public/scripts/openai.js (hunk 2 above)
Comment (ZhenyaPav) [reply public/scripts/openai.js]: Done

File: public/scripts/openai.js [hunk 8]
```diff
@@ -885,6 +895,10 @@ async function populateChatHistory(messages, prompts, chatCompletion, type = nul
     const audioInlining = isAudioInliningSupported();
     const canUseTools = ToolManager.isToolCallingSupported();
     const includeSignature = isReasoningSignatureSupported();
+    const isOpenRouterSource = oai_settings.chat_completion_source === chat_completion_sources.OPENROUTER;
+    const toolReasoningMode = isOpenRouterSource ? getOpenRouterToolReasoningMode() : openrouter_tool_reasoning_modes.ACTIVE_CHAIN;
```
Comment (Cohee1207) [inline public/scripts/openai.js]: Why is it set to `ACTIVE_CHAIN` for non-OpenRouter sources?

File: public/scripts/openai.js [hunk 9]
```diff
@@ -936,11 +950,63 @@ async function populateChatHistory(messages, prompts, chatCompletion, type = nul
         }
 
         if (canUseTools && Array.isArray(chatPrompt.invocations)) {
+            const promptIdx = messages.indexOf(chatPrompt);
+            const reasoningIsEligible = toolReasoningMode !== openrouter_tool_reasoning_modes.DISABLED
```
Comment (Cohee1207) [inline public/scripts/openai.js]: Looks like this will result in adding reasoning to messages that are on non-OpenRouter sources.

File: public/scripts/openai.js [hunk 10]
```diff
@@ -936,11 +950,63 @@ async function populateChatHistory(messages, prompts, chatCompletion, type = nul
...
+                } else if (previousAssistantReasoning) {
+                    // Prefer currently editable assistant-text reasoning based on forwarding mode over invocation snapshot.
+                    clone.reasoning = previousAssistantReasoning;
+                }
+                return clone;
+            });
             const toolCallMessage = await Message.createAsync(chatMessage.role, undefined, 'toolCall-' + chatMessage.identifier);
             const toolResultMessages = await Promise.all(invocations.slice().reverse().map((invocation) => Message.createAsync('tool', invocation.result || '[No content]', invocation.id)));
-            await toolCallMessage.setToolCalls(invocations, includeSignature);
+            await toolCallMessage.setToolCalls(invocations, includeSignature, shouldForwardOpenRouterReasoning);
```
Comment (Cohee1207) [inline public/scripts/openai.js]: Let's call it just "includeToolReasoning" without mentioning OpenRouter, anticipating potential improvements for other sources.

**Comment by ZhenyaPav** [comment]
Reworked the logic a bit:

- Made a list of eligible providers (for now, only OpenRouter)
- If our provider is in the list, use the dropdown value, otherwise DISABLED
- Variables were also renamed to remove references to openrouter

The previous implementation should not have resulted in non-OR providers getting interleaved reasoning in the requests, but this one is cleaner and easier to expand (as far as I understand, we'll really only need to add the provider to the list, and to data-source for the UI dropdown, assuming the provider expects the same implementation as OR)

File: public/scripts/openai.js [hunk 11]
```diff
@@ -936,11 +957,63 @@ async function populateChatHistory(messages, prompts, chatCompletion, type = nul
         }
 
         if (canUseTools && Array.isArray(chatPrompt.invocations)) {
+            const promptIdx = messages.indexOf(chatPrompt);
+            const reasoningIsEligible = toolReasoningMode !== tool_reasoning_modes.DISABLED
+                && promptIdx > lastUserIdx;
```
Comment (Cohee1207) [inline public/scripts/openai.js:962]: Just verifying.. What if the prompt doesn't have user messages (i.e. lastUserIdx === -1)?

**Review by Cohee1207 — APPROVED** [review]
Left one question. The rest of the code is ok. Will proceed with testing soon-ish. If you can find someone to test besides you and me, that'll be even better - more edge cases could be found.

File: public/scripts/openai.js (hunk 11 above)
Comment (ZhenyaPav) [reply public/scripts/openai.js:962]: Should not result in any issues. If the prompt doesn't have user messages (character messages and tool only), then reasoningIsEligible will be true for all of them.

I tested this case just now, the app behaves as expected - no issues, reasoning is included with every message from the start.

Comment (Cohee1207) [reply public/scripts/openai.js:962]: Thanks for confirming. 🙏

**Comment by Cohee1207** [comment]
Tested on Gemini 3, Sonnet 4.6, GLM 5, Kimi K2.5, MiniMax M2.5 via OpenRouter. No explicit provider selection, enabled thinking with minimal effort.

Prompt: "Generate an image that reflects the current news."
Tools used: built-in image generation tool and [RSS extension](https://github.com/SillyTavern/Extension-RSS).

All 3 modes on all specified models: no issues.

Didn't found a significant difference/improvement in quality of responses either, but I guess a more sophisticated workflow can benefit from that better.

**Comment by Cohee1207** [comment]
@codex review

File: public/script.js [hunk 12]
```diff
@@ -5341,7 +5343,7 @@ export async function Generate(type, { automatic_trigger, force_name2, quiet_pro
             const hasToolCalls = ToolManager.hasToolCalls(data);
             const shouldDeleteMessage = type !== 'swipe' && ['', '...'].includes(getMessage) && !reasoning;
             hasToolCalls && shouldDeleteMessage && await deleteLastMessage();
-            const invocationResult = await ToolManager.invokeFunctionTools(data);
+            const invocationResult = await ToolManager.invokeFunctionTools(data, { reasoningText: reasoning });
```
[the remark this answers is not in the document]
Comment (Cohee1207) [reply public/script.js:5346]: I think interleaved should also default to DISABLED if the "Request reasoning" toggle is off? Otherwise it makes no sense.

File: public/scripts/openai.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (Cohee1207) [reply public/scripts/openai.js:962]: Idk if it's really an issue. Continue on Chat Completion is a meme anyway.

**Comment by ZhenyaPav** [comment]
I implemented a guard so interleaved thinking only runs when `Request model reasoning` is enabled, and a hint is shown below interleaved thinking if request reasoning is disabled
